File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointList/mountpoint/MountPoint.java
package ru.nsu.bobrofon.easysshfs.mountpointList.mountpoint;

import android.content.Context;
import android.database.Observable;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.util.Log;
import android.util.Pair;

import com.topjohnwu.superuser.Shell;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.LinkedList;
import java.util.List;

import ru.nsu.bobrofon.easysshfs.EasySSHFSActivity;
import ru.nsu.bobrofon.easysshfs.log.LogSingleton;

public class MountPoint {
	private String mPointName;
	private boolean mAutoMount;
	private String mUserName;
	private String mHost;
	private int mPort;
	private String mPassword;
	private boolean mStorePassword;
	private String mRemotePath;
	private String mLocalPath;
	private boolean mForcePermissions;
	private String mOptions;
	private String mRootDir;

	private MountObservable mObservable = new MountObservable();
	private boolean mIsMounted = false;

	private final static String DEFAULT_OPTIONS =
		"password_stdin,"
		+ "UserKnownHostsFile=/dev/null,"
		+ "StrictHostKeyChecking=no,"
		+ "rw,"
		+ "dirsync,"
		+ "nosuid,"
		+ "nodev,"
		+ "noexec,"
		+ "umask=0,"
		+ "allow_other,"
		+ "uid=9997,"
		+ "gid=9997";

	public MountPoint() {
		mPointName = "";
		mAutoMount = false;
		mUserName = "";
		mHost = "";
		mPort = 22;
		mPassword = "";
		mStorePassword = false;
		mRemotePath = "";
		mLocalPath = "";
		mForcePermissions = false;
		mOptions = DEFAULT_OPTIONS;
		mRootDir = "";
	}

	public void setPointName(final String name) {
		mPointName = name;
	}

	public void setAutoMount(final boolean autoMount) {
		mAutoMount = autoMount;
	}

	public void setUserName(final String userName) {
		mUserName = userName;
	}

	public void setHost(final String host) {
		mHost = host;
	}

	public void setPort(final int port) {
		if (port < 0 || port > 65535) {
			logMessage("Port " + port + " is out of range [0; 65535]");
			return;
		}
		mPort = port;
	}

	public void setPort(final String port) {
		try {
			setPort(Integer.parseInt(port));
		}
		catch (final NumberFormatException e) {
			logMessage("'" + port + "' is invalid port value");
		}
	}

	public void setPassword(final String password) {
		mPassword = password;
	}

	public void setStorePassword(final boolean storePassword) {
		mStorePassword = storePassword;
	}

	public void setRemotePath(final String remotePath) {
		mRemotePath = remotePath;
	}

	public void setLocalPath(final String localPath) {
		mLocalPath = localPath;
	}

	public void setForcePermissions(final boolean forcePermissions) {
		mForcePermissions = forcePermissions;
	}

	public void setOptions(final String options) {
		mOptions = options;
	}

	public void setRootDir(final String rootDir) {
		mRootDir = rootDir;
	}

	public String getPointName() {
		if (!mPointName.isEmpty()) {
			return mPointName;
		}
		else {
			return mLocalPath;
		}
	}

	public boolean getAutoMount() {
		return mAutoMount;
	}

	public String getUserName() {
		return mUserName;
	}

	public String getHost() {
		return mHost;
	}

	public int getPort() {
		return mPort;
	}

	public String getPassword() {
		return mPassword;
	}

	public boolean getStorePassword() {
		return mStorePassword;
	}

	public String getRemotePath() {
		return mRemotePath;
	}

	public String getLocalPath() {
		return mLocalPath;
	}

	public boolean getForcePermissions() {
		return mForcePermissions;
	}

	public String getOptions() {
		return mOptions;
	}

	public JSONObject json() {
		JSONObject selfJson = new JSONObject();
		try {
			selfJson.put("PointName", mPointName);
			selfJson.put("AutoMount", mAutoMount);
			selfJson.put("UserName", mUserName);
			selfJson.put("Host", mHost);
			selfJson.put("Port", mPort);
			if (mStorePassword) {
				selfJson.put("Password", mPassword);
			}
			selfJson.put("RemotePath", mRemotePath);
			selfJson.put("LocalPath", mLocalPath);
			selfJson.put("ForcePermissions", mForcePermissions);
			selfJson.put("Options", mOptions);
			selfJson.put("RootDir", mRootDir);
		} catch (JSONException e) {
			String TAG = "MOUNT_POINT";
			Log.e(TAG, e.getMessage());
		}

		return selfJson;
	}

	public void json(JSONObject selfJson) {
		mPointName = selfJson.optString("PointName", mPointName);
		mAutoMount = selfJson.optBoolean("AutoMount", mAutoMount);
		mUserName = selfJson.optString("UserName", mUserName);
		mHost = selfJson.optString("Host", mHost);
		mPort = selfJson.optInt("Port", mPort);
		mStorePassword = selfJson.optBoolean("StorePassword", mStorePassword);
		if (selfJson.has("Password")) {
			mStorePassword = true;
			mPassword = selfJson.optString("Password", mPassword);
		}
		else {
			mStorePassword = false;
			mPassword = "";
		}
		mRemotePath = selfJson.optString("RemotePath", mRemotePath);
		mLocalPath = selfJson.optString("LocalPath", mLocalPath);
		mForcePermissions = selfJson.optBoolean("ForcePermissions", mForcePermissions);
		mOptions = selfJson.optString("Options", mOptions);
		mRootDir = selfJson.optString("RootDir", mRootDir);
	}

	private String getHostIp() {
		try {
			InetAddress address = InetAddress.getByName(getHost());
			if (address instanceof Inet6Address) {
				if (!address.getHostAddress().startsWith("[")) {
					return '[' + address.getHostAddress() + ']';
				}
			}
			return address.getHostAddress();

		} catch (UnknownHostException e) {
			logMessage(e.getMessage());
			return getHost();
		}
	}

	private void logMessage(final String message) {
		LogSingleton.getLogModel().addMessage(message);
	}

	public boolean isMounted() {
		return mIsMounted;
	}

	public void checkMount(final Context context) {
		checkMount(false, context);
	}

	private void checkMount(final boolean verbose, final Context context) {
		new CheckMountTask(verbose, context).execute();
	}

	public void mount(final Context context, final Shell shell) {
		mount(false, context, shell);
	}

	public void mount(boolean verbose, final Context context, final Shell shell) {
		logMessage("mount");
		new MountTask(verbose, context, shell).execute();
	}

	public void umount(final Context context, final Shell shell) {
		umount(false, context, shell);
	}

	public void umount(final boolean verbose, final Context context, final Shell shell) {
		String umountCommand = "umount ";
		if (isBusyboxAvailable(shell)) {
			umountCommand = "busybox umount -f ";
		}
		logMessage(umountCommand);
		runCommand(umountCommand + getLocalPath(), verbose, context, shell);
	}

	public void registerObserver(final Observer observer, final Context context) {
		mObservable.registerObserver(observer);
		checkMount(context);
	}

	public void unregisterObserver(final Observer observer) {
		mObservable.unregisterObserver(observer);
	}

	public interface Observer {
		void onMountStateChanged(final MountPoint mountPoint);
	}

	private class MountObservable extends Observable<Observer> {
		public void notifyChanged() {
			for (final Observer observer : mObservers) {
				observer.onMountStateChanged(MountPoint.this);
			}
		}
	}

	private class CheckMountTask extends AsyncTask<Void, Void, Pair<Boolean, String>> {
		private final String mMountFile = "/proc/mounts";
		private final boolean mVerbose;
		private final Context mContext;

		private CheckMountTask(final boolean verbose, final Context context) {
			this.mVerbose = verbose;
			this.mContext = context;
		}

		@Override
		protected Pair<Boolean, String> doInBackground(Void... params) {
			final StringBuilder mountLine = new StringBuilder();
			mountLine.append(getUserName()).append('@');
			mountLine.append(getHostIp()).append(':');

			String canonicalLocalPath = getLocalPath();
			try {
				canonicalLocalPath = new File(canonicalLocalPath).getCanonicalPath();
			} catch (IOException e) {
				logMessage("Can't get canonical path of " + getLocalPath() + " : " + e.getMessage());
			}

			mountLine.append(getRemotePath()).append(' ').append(canonicalLocalPath);
			mountLine.append(' ').append("fuse.sshfs").append(' ');

			boolean result = false;

			try {
				FileInputStream fstream = new FileInputStream(mMountFile);
				BufferedReader br = new BufferedReader(new InputStreamReader(fstream));

				String line;
				while (!result && (line = br.readLine()) != null) {
					result = line.contains(mountLine.toString());
				}

				br.close();
			} catch (FileNotFoundException e) {
				return new Pair<>(result, e.getMessage());
			} catch (IOException e) {
				return new Pair<>(result, e.getMessage());
			}

			mIsMounted = result;
			if (result) {
				return new Pair<>(result, "Pattern " + mountLine.toString() + " is in " + mMountFile);
			}
			else {
				return new Pair<>(result, "Pattern " + mountLine.toString() + " is not in " + mMountFile);
			}
		}

		@Override
		protected void onPostExecute(final Pair<Boolean, String> result) {
			mIsMounted = result.first;
			logMessage(result.second);

			mObservable.notifyChanged();
			if (mVerbose) {
				EasySSHFSActivity.showToast("done", mContext);
			}
		}
	}

	private class MountTask extends AsyncTask<Void, Void, String> {
		private final boolean mVerbose;
		private final Context mContext;
		private final Shell mShell;

		private MountTask(boolean verbose, final Context context, final Shell shell) {
			this.mVerbose = verbose;
			this.mContext = context;
			this.mShell = shell;
		}

		@Override
		protected String doInBackground(Void... params) {
			return getHostIp();
		}

		@Override
		protected void onPostExecute(final String hostIp) {
			String command = fixLocalPath() + "echo '" + getPassword() + "' | " +
				mRootDir + "/sshfs" +
				" -o 'ssh_command=" + mRootDir + "/ssh" + ',' +
				getOptions() + ",port=" + getPort() + "' " +
				getUserName() + '@' + hostIp + ':' +
				getRemotePath() + ' ' + getLocalPath();

			runCommand(command, mVerbose, mContext, mShell);
		}

		private String fixLocalPath() {
			if (!getForcePermissions()) {
				return "";
			}
			return "mkdir -p " + getLocalPath() + " ; " +
				"chmod 777 " + getLocalPath() + " ; " +
				"chown 9997:9997 " + getLocalPath() + " ; ";
		}
	}

	private void runCommand(final String command, final boolean verbose, final Context aContext, final Shell shell) {
		final List<String> stdout = new LinkedList<>();
		final List<String> stderr = new LinkedList<>();
		shell.run(stdout, stderr, new Shell.Async.Callback() {
			@Override
			public void onTaskResult(@Nullable List<String> stdout, @Nullable List<String> stderr) {
				logAll(stdout);
				logAll(stderr);
				checkMount(verbose, aContext);
			}

			private void logAll(@Nullable final List<String> stdio) {
				if (stdio == null) {
					return;
				}
				for (final String line: stdio) {
					logMessage(line);
				}
			}
		}, command);
	}

	private boolean isBusyboxAvailable(final Shell _shell) {
		return _shell.testCmd("busybox");
	}

}

File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointlist/mountpoint/MountPoint.kt
// SPDX-License-Identifier: MIT
package ru.nsu.bobrofon.easysshfs.mountpointlist.mountpoint

import android.content.Context
import android.os.AsyncTask
import android.util.Log
import android.util.Pair
import com.topjohnwu.superuser.Shell
import org.json.JSONException
import org.json.JSONObject
import ru.nsu.bobrofon.easysshfs.EasySSHFSActivity
import ru.nsu.bobrofon.easysshfs.IpResolver
import ru.nsu.bobrofon.easysshfs.log.AppLog
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.lang.ref.WeakReference
import java.util.*


class MountPoint(
    var pointName: String = "",
    var autoMount: Boolean = false,
    var userName: String = "",
    var host: String = "",
    port: Int = 22,
    var password: String = "",
    var storePassword: Boolean = false,
    var remotePath: String = "",
    var localPath: String = "",
    var forcePermissions: Boolean = false,
    var options: String = DEFAULT_OPTIONS,
    var rootDir: String = "",
    var identityFile: String = "",
    isMounted: Boolean = false,
    private val appLog: AppLog = AppLog.instance()
) {
    private val observable = MountPointObservable()

    var isMounted = isMounted
        private set

    val visiblePointName: String
        get() = pointName.ifEmpty { localPath }

    var port: Int = port
        set(value) {
            if (value < 0 || value > 65535) {
                logMessage("Port $value is out of range [0; 65535]")
                return
            }
            field = value
        }

    private fun hostIp(foreground: Boolean = true) = IpResolver.resolve(host, foreground)

    fun setPort(value: String) {
        try {
            port = Integer.parseInt(value)
        } catch (e: NumberFormatException) {
            logMessage("'$port' is invalid port value")
        }

    }

    fun json(): JSONObject {
        val selfJson = JSONObject()
        try {
            selfJson.put("PointName", pointName)
            selfJson.put("AutoMount", autoMount)
            selfJson.put("UserName", userName)
            selfJson.put("Host", host)
            selfJson.put("Port", port)
            if (storePassword) {
                selfJson.put("Password", password)
            }
            selfJson.put("RemotePath", remotePath)
            selfJson.put("LocalPath", localPath)
            selfJson.put("ForcePermissions", forcePermissions)
            selfJson.put("Options", options)
            selfJson.put("RootDir", rootDir)
            selfJson.put("IdentityFile", identityFile)
        } catch (e: JSONException) {
            Log.e(TAG, e.message ?: "")
        }

        return selfJson
    }

    fun json(selfJson: JSONObject) {
        pointName = selfJson.optString("PointName", pointName)
        autoMount = selfJson.optBoolean("AutoMount", autoMount)
        userName = selfJson.optString("UserName", userName)
        host = selfJson.optString("Host", host)
        port = selfJson.optInt("Port", port)
        storePassword = selfJson.optBoolean("StorePassword", storePassword)
        if (selfJson.has("Password")) {
            storePassword = true
            password = selfJson.optString("Password", password)
        } else {
            storePassword = false
            password = ""
        }
        remotePath = selfJson.optString("RemotePath", remotePath)
        localPath = selfJson.optString("LocalPath", localPath)
        forcePermissions = selfJson.optBoolean("ForcePermissions", forcePermissions)
        options = selfJson.optString("Options", options)
        rootDir = selfJson.optString("RootDir", rootDir)
        identityFile = selfJson.optString("IdentityFile", identityFile)
    }

    private fun logMessage(message: String) {
        appLog.addMessage(message)
    }

    private fun checkMount(context: Context? = null) {
        CheckMountTask(this, context).execute()
    }

    fun mount(shell: Shell, context: Context? = null) {
        logMessage("mount")
        MountTask(this, shell, context).execute()
    }

    fun umount(shell: Shell, context: Context? = null) {
        if (context == null && !isMounted) {
            Log.d(TAG, "skip umount because we are in the background and it is not mounted")
            return
        }

        val umountCommand = "umount -l -v $localPath"
        logMessage(umountCommand)
        runCommand(umountCommand, shell, context)
    }

    fun registerObserver(observer: MountStateChangeObserver) {
        observable.registerObserver(observer)
        checkMount()
    }

    fun unregisterObserver(observer: MountStateChangeObserver) {
        observable.unregisterObserver(observer)
    }

    private fun runCommand(command: String, shell: Shell, context: Context? = null) {
        Log.i("shell", command)
        shell.newJob().add(command).to(LinkedList(), LinkedList()).submit { result ->
            logAll(result.out)
            logAll(result.err)
            logMessage("exit code: ${result.code}")
            checkMount(context)
        }
    }

    private fun logAll(stdio: List<String>?) {
        if (stdio == null) {
            return
        }
        for (line in stdio) {
            logMessage(line)
        }
    }

    companion object {

        private const val DEFAULT_OPTIONS = (
                "password_stdin,"
                        + "UserKnownHostsFile=/dev/null,"
                        + "StrictHostKeyChecking=no,"
                        + "rw,"
                        + "dirsync,"
                        + "nosuid,"
                        + "nodev,"
                        + "noexec,"
                        + "umask=0,"
                        + "allow_other,"
                        + "uid=9997,"
                        + "gid=9997")

        private const val TAG = "MOUNT_POINT"

        private class CheckMountTask constructor(
            private val mountPoint: MountPoint,
            private val context: WeakReference<Context?> = WeakReference(null)
        ) :
            AsyncTask<Void, Void, Pair<Boolean?, String>>() {

            private val mountFile = "/proc/mounts"

            constructor(mountPoint: MountPoint, context: Context?) :
                    this(mountPoint, WeakReference(context))

            @Deprecated("Deprecated in Java")
            override fun doInBackground(vararg params: Void): Pair<Boolean?, String> {
                val fromForeground = context.get() != null
                val mountLine = StringBuilder()
                mountLine.append("${mountPoint.userName}@${mountPoint.hostIp(fromForeground)}:")

                val canonicalLocalPath: String
                try {
                    canonicalLocalPath = File(mountPoint.localPath).canonicalPath
                } catch (e: IOException) {
                    return Pair(
                        null,
                        "Can't get canonical path of ${mountPoint.localPath}: ${e.message}"
                    )
                }

                mountLine.append("${mountPoint.remotePath} $canonicalLocalPath fuse.sshfs ")

                val result: Boolean
                try {
                    val procmount = File(mountFile)
                    result = procmount.useLines { lines ->
                        lines.any { line -> line.contains(mountLine.toString()) }
                    }
                } catch (e: FileNotFoundException) {
                    return Pair(null, e.message ?: "")
                } catch (e: IOException) {
                    return Pair(null, e.message ?: "")
                }

                return if (result) {
                    Pair(true, "Pattern $mountLine is in $mountFile")
                } else {
                    Pair(false, "Pattern $mountLine is not in $mountFile")
                }
            }

            @Deprecated("Deprecated in Java")
            override fun onPostExecute(result: Pair<Boolean?, String>) {
                mountPoint.isMounted = result.first ?: mountPoint.isMounted
                mountPoint.logMessage(result.second)

                mountPoint.observable.notifyChanged()
                context.get()?.let {
                    EasySSHFSActivity.showToast(
                        "done",
                        it
                    )
                }
            }
        }

        private class MountTask constructor(
            private val mountPoint: MountPoint,
            private val shell: Shell,
            private val context: WeakReference<Context?> = WeakReference(null)
        ) : AsyncTask<Void, Void, String>() {

            constructor(mountPoint: MountPoint, shell: Shell, context: Context?) :
                    this(mountPoint, shell, WeakReference(context))

            @Deprecated("Deprecated in Java")
            override fun doInBackground(vararg params: Void): String {
                val fromForeground = context.get() != null
                return mountPoint.hostIp(fromForeground)
            }

            @Deprecated("Deprecated in Java")
            override fun onPostExecute(hostIp: String) {
                val command = "${ensureLocalPath()} echo '${mountPoint.password}' | " +
                        "${mountPoint.rootDir}/sshfs" +
                        " -o 'ssh_command=${mountPoint.rootDir}/ssh," +
                        "${mountPoint.options}${identity()},port=${mountPoint.port}' " +
                        "${mountPoint.userName}@$hostIp:" +
                        "${mountPoint.remotePath} ${mountPoint.localPath}"

                mountPoint.runCommand(command, shell, context.get())
            }

            private fun identity(): String {
                return if (mountPoint.identityFile.isEmpty()) {
                    ""
                } else ",IdentityFile=${mountPoint.identityFile}"
            }

            private fun ensureLocalPath(): String {
                if (context.get() == null) {
                    return "ls ${mountPoint.localPath} && "
                }
                if (mountPoint.forcePermissions) {
                    return "mkdir -p ${mountPoint.localPath} ; " +
                            "chmod 777 ${mountPoint.localPath} ; " +
                            "chown 9997:9997 ${mountPoint.localPath} ; "
                }
                return ""
            }
        }
    }
}
